RubberduckTests/Rewriter/MemberAttributeRecovererTests.cs

```diff
@@ -0,0 +1,265 @@
+                var declarationsForWhichToRemoveAttributes = state.DeclarationFinder.UserDeclarations(DeclarationType.Function);
+                RemoveAttributes(declarationsForWhichToRemoveAttributes, rewriteSession);
+
+                ExecuteAndWaitForParserState(state, () => rewriteSession.TryRewrite(), ParserState.Ready);
+            }
+            var actualCode = component.CodeModule.Content();
+            Assert.AreEqual(expectedCode, actualCode);
+        }
+
+        private static void ExecuteAndWaitForParserState(RubberduckParserState state, Action action, ParserState parserStateToAwait)
+        {
+            using (var waitHandle = new EventWaitHandle(false, EventResetMode.ManualReset))
+            {
+                var parserStateAwaiter = new ParserStateAwaiter(waitHandle, new List<ParserState> { parserStateToAwait });
+                state.StateChanged += parserStateAwaiter.ParserStateHandler;
+
+                action.Invoke();
+
+                waitHandle.WaitOne();
+                state.StateChanged -= parserStateAwaiter.ParserStateHandler;
+            }
+        }
+
+        private static void RemoveAttributes(IEnumerable<Declaration> declarationsForWhichToRemoveAttributes, IRewriteSession rewriteSession)
+        {
+            foreach (var declaration in declarationsForWhichToRemoveAttributes)
+            {
+                foreach (var attribute in declaration.Attributes)
+                {
+                    //We cannot remove use the attributesUpdater here because it requires an attribute rewrite session,
+                    //which rewrites in a suspended state. 
+                    var attributeContext = attribute.Context;
+                    var rewriter = rewriteSession.CheckOutModuleRewriter(declaration.QualifiedModuleName);
+                    rewriter.Remove(attribute.Context);
+                    if (attributeContext.TryGetFollowingContext(out VBAParser.EndOfLineContext followingEndOfLine))
+                    {
+                        rewriter.Remove(followingEndOfLine);
+                    }
+                }
+            }
+        }
+
+        [Test]
+        [Category("Rewriter")]
+        public void RecoveringAttributesRecoversTheAttributesInTheModulesProvided_ViaModule()
+        {
+            var inputCodeWithRecovery =
+                @"Public Function Foo() As Variant
+Attribute Foo.VB_UserMemId = 0
+Attribute Foo.VB_Description = ""DefaultMember""
+End Function
+
+Public Function Bar() As Variant
+Attribute Bar.VB_UserMemId = -4
+Attribute Bar.VB_Description = ""Enumerator""
+End Function";
+
+            var inputCodeWithoutRecovery =
+                @"Public Function Foo() As Variant
+Attribute Foo.VB_UserMemId = 0
+Attribute Foo.VB_Description = ""DefaultMember""
+End Function
+
+Public Function Bar() As Variant
+Attribute Bar.VB_UserMemId = -4
+Attribute Bar.VB_Description = ""Enumerator""
+End Function";
```

Comment:
Why not just `var inputCode = @"...";`?

Reply:
But if both components should be using the same input code, then that would be made explicit if they both used the same `inputCode`, rather than identical copies... no?

Reply:
Sorry, I did not get that this is what your first comment was trying to say.